### ISSUE TITLE
Migrate from setuptools/tox to hatch

### ISSUE THREAD

**Issue by jcgruenhage** [latchset/jwcrypto#304]
More and more packages are migrating to more modern build/env tooling, and one of those tools is hatch. It can replace both setuptools and tox here, and it's also the one used in the official PyPI guides. Interestingly, even tox itself is migrating to hatch [for their rewrite](https://github.com/tox-dev/tox/blob/rewrite/pyproject.toml#L3)

Before I invest the time of creating a PR for this, would this be something you're interested in?

**Comment by ofek** [comment]
Should be easy with https://hatch.pypa.io/latest/intro/#existing-project

**Comment by jcgruenhage** [comment]
Yeah, that takes care of the setuptools part, but it still requires some manual work (did it out of curiosity and the version detection and license name both needed some adjustment) and doesn't take care of the bigger/more complicated task of migrating the tox envs into hatch as well.

**Comment by jcgruenhage** [comment]
Oh, and not to mention migrating the stuff from the Makefile into hatch as well.

**Comment by simo5** [comment]
Ok I do not know hatch, so I would like to know a few things:
- What are the advantages of migrating
- What are the disadvantages of migrating
- What is the maturity level of hatch
- What platforms would I leave behind migrating to hatch

**Comment by jcgruenhage** [comment]
For some of those questions, @ofek is surely better qualified to answer, but considering I started this I'll answer to the best of my knowledge and @ofek can jump in if I get anything wrong.

- **advantages of migrating**: I find hatch nicer to use than the plethora of tools it replaces. Having one tool with all the batteries included, like it's common for other languages (think npm and cargo for example), makes working on repositories quite a bit easier IMO. But that's kinda personal preference. In general, as mentioned in the OP, hatch is tooling by the same group as setuptools (namely the Python Packaging Authority), but it's just a way more modern piece of software, and as such is being used in examples on packaging.python.org all over the place.
- **disadvantages of migrating**: Contributors (and maintainers) need to adjust their workflows accordingly. Instead of `make`, `tox`, `python setup.py` etc, they'll have to run the hatch equivalent. Can't really think of other disadvantages here. My interactions with hatch (both from my perspective as someone packaging python software for a Linux distribution, as well as from the perspective of a library developer publishing software to pypi.org and running tests, linting etc) have been extremely smooth.
- **maturiry**: As an official PyPA project, which is following semver and has had a 1.0.0 release in the past, I think it can be considered production ready. https://github.com/pypa/hatch/blob/master/pyproject.toml#L26 agrees with that, and the adoption over at https://hatch.pypa.io/latest/users/ kinda speaks for itself as well. Quite a few big projects are adopting it, and again, as already mentioned in the OP, even tox is migrating to it for their rewrite, at least on the build backend side of things.
- **platforms left behind**:  I think for developers of jwcrypto, this means that python 3.6 can't be used anymore, I *think* for users it doesn't change anything, but I'm not 100% sure on that. 3.6 is EOL anyway though, so I think that doesn't matter much.

Anything I forgot to mention, @ofek?

**Comment by ofek** [comment]
> What platforms would I leave behind migrating to hatch

Hatchling is available on all the major distribution channels such as [Debian](https://salsa.debian.org/python-team/packages/hatchling), [Fedora](https://src.fedoraproject.org/rpms/python-hatchling), [Arch Linux](https://archlinux.org/packages/community/any/python-hatchling/), [conda-forge](https://anaconda.org/conda-forge/hatchling), [Nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/hatchling/default.nix), [Alpine Linux](https://github.com/alpinelinux/aports/blob/master/community/py3-hatchling/APKBUILD), [FreeBSD](https://github.com/freebsd/freebsd-ports/blob/main/devel/py-hatchling/Makefile)/[OpenBSD](https://github.com/openbsd/ports/blob/master/devel/py-hatchling/Makefile), [Gentoo Linux](https://github.com/gentoo/gentoo/tree/master/dev-python/hatchling), [MacPorts](https://github.com/macports/macports-ports/blob/master/python/py-hatchling/Portfile), [OpenEmbedded](https://github.com/openembedded/openembedded-core/commit/846e806181f1349be29cbce78c5041735dfd7e6f), [Spack](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-hatchling/package.py), [MSYS2](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-python-hatchling/PKGBUILD), etc.

**Comment by simo5** [comment]
Considering I recently change the code to:  python_requires = '>= 3.6' it sounds like not a huge deal to move to >= 3.7

If this is "the way of the future", who am I to stand in the way ?

I would definitely review a PR.

As for the Makefile, it would be nice to preserve the targets that can be preserved, it is my way to re-learn how to run the various python tools (in this case will be how to use hatch) when I stay a way for long enough to forget.

**Comment by jcgruenhage** [comment]
> Considering I recently change the code to: python_requires = '>= 3.6' it sounds like not a huge deal to move to >= 3.7

I'm not *sure* this is necessary for users installing the library, just for building it, but maybe @ofek can clarify this bis? 

> As for the Makefile, it would be nice to preserve the targets that can be preserved, it is my way to re-learn how to run the various python tools (in this case will be how to use hatch) when I stay a way for long enough to forget.

So that'd be a Makefile wrapping hatch basically? Because I'd prefer to have the commands for testing, linting etc in a document instead, but I can surely also add a Makefile if you want it.

**Comment by simo5** [comment]
Yes the Makefile is just a muscle-memory convenience, I am totally for a Doc as well.

**Comment by ofek** [comment]
> I'm not sure this is necessary for users installing the library, just for building it

Correct

**Comment by ofek** [comment]
~3-5% https://pypistats.org/packages/jwcrypto

**Comment by simo5** [comment]
Cool stats, seem lion share is 3.7 in terms of downloads, but that counts only pypi.

**Comment by ofek** [comment]
Such environments are unlikely to be using new versions of libraries anyway so they'd never hit this

**Comment by ofek** [comment]
Any update on this?

**Comment by jcgruenhage** [comment]
I haven't been able to work on this yet, because I've had other more urgent stuff to work on. I still plan on working on this though, unless someone else is quicker of course. I'll update this issue when I actually start on it :)
